substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1050,17 +1117,35 @@ impl<T: Config> Pallet<T> {
 	/// `Validators` or `VoterList` outside of this function is almost certainly
 	/// wrong.
 	pub fn do_remove_validator(who: &T::AccountId) -> bool {
-		let outcome = if Validators::<T>::contains_key(who) {
-			Validators::<T>::remove(who);
-			let _ = T::VoterList::on_remove(who).defensive();
-			true
-		} else {
-			false
+		let outcome = match Self::status(who) {
+			Ok(StakerStatus::Validator) => {
+				// make sure the validator is chilled before removing it.
+				let outcome = Self::do_chill_validator(who);
+				T::EventListeners::on_validator_remove(who);
+				outcome
+			},
+			Ok(StakerStatus::Nominator(_)) => {
+				defensive!("called do_validator_remove on a nominator stash.");
+				false
+			},
+			Ok(StakerStatus::Idle) | Err(_) if T::TargetList::contains(who) => {
```

Comment:
iterates over TargetList again. i'm sure this is safe but just flagging as I don't understand why

Reply:
I don't think `Err(_)` belongs here because in the `stake-tracker` impl of `on_validator_remove`, you do the `StakerStatus` check again and you correctly point out that for an `Err` value, the target is non-existent.

To add to what @liamaharon was saying, we shouldn't do this check (and iteration) twice. We should choose one entity that is responsible for this and my gut feeling tells me it's the `on_validator_remove` impl.